### PR TITLE
Add palette encoding to libdvi, plus example

### PR DESCRIPTION
**Pull request by MichaelBell**
The palette encoder currently allows a 64 colour palette, where each entry is a 16-bit colour (could easily be expanded to 18-bit, or full 24-bit if we include more symbols).

You first call `tmds_setup_palette_symbols` with you colour palette to make the TMDS symbol arrays, and then `tmds_encode_palette_data` produces a full 3 channel scanline for queueing.

The assembly encode loop is a very tiny tweak to the standard full res encode loop, but now only reading one packed word per 4 output symbols as it is turning 8-bit palette entries into symbols, rather than 16-bit colour values.

### CODE REVIEW

**Comment by MichaelBell** [comment]
I've now generalised this to allow 256 colour palettes, at the same performance (although we could now do 64 colour palettes at better performance).

So you might want to ignore this pull request until I can update with that (should be tomorrow).

**Comment by MichaelBell** [comment]
Palette encoder now allows 256 colours in the palette, and works with different sized colour palettes.

File: software/libdvi/tmds_encode.S [hunk 1]
```diff
@@ -395,6 +395,10 @@ tmds_2bpp_table:
 	// DC balance defined to be 0 at start of scanline:
 	movs r4, #0
 	str r4, [r2, #ACCUM1_OFFS]
+#if TMDS_FULLRES_NO_DC_BALANCE
```
Comment (Wren6991) [inline software/libdvi/tmds_encode.S:398]: oops :)

File: software/libdvi/tmds_encode.c [hunk 2]
```diff
@@ -162,3 +162,70 @@ void __not_in_flash_func(tmds_encode_data_channel_fullres_16bpp)(const uint32_t
...
+	uint core = get_core_num();
+#if !TMDS_FULLRES_NO_INTERP_SAVE
+	interp_hw_save_t interp0_save, interp1_save;
+	interp_save(interp0_hw, &interp0_save);
+	interp_save(interp1_hw, &interp1_save);
+#endif
+
+	interp0_hw->base[2] = (uint32_t)tmds_palette;
+	interp1_hw->base[2] = (uint32_t)tmds_palette;
+	interp0_hw->ctrl[0] = 0x40u | ((palette_bits + 1) << SIO_INTERP0_CTRL_LANE0_MASK_MSB_LSB);
```
Comment (Wren6991) [inline software/libdvi/tmds_encode.c:201]: Could you clean up the magic numbers please (in return I will clean up the ones I lazily left in vista/main.c)

File: software/libdvi/tmds_encode.S [hunk 3]
```diff
@@ -529,3 +533,118 @@ decl_func_x tmds_fullres_encode_loop_16bpp_leftshift_x
...
 decl_func_y tmds_fullres_encode_loop_16bpp_leftshift_y
 	tmds_fullres_encode_loop_16bpp_leftshift
+
+
+// Variant of tmds_fullres_encode_loop_16bpp that reads
+// 8-bit wide pixels packed 4 per word.  The interpolator
+// base is set to a reordered list of TMDS symbols based
+// on a user colour palette.
+
+.macro tmds_palette_encode_loop
```
Comment (Wren6991) [inline software/libdvi/tmds_encode.S:543]: Very nice! I might come through and add some macro goodness here later, along with the other fullres loops.

File: software/apps/mandel-full/main.c [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+#include "mandelbrot.h"
+
+// TMDS bit clock 252 MHz
+// DVDD 1.2V (1.1V seems ok too)
+#define FRAME_WIDTH 640
+#define FRAME_HEIGHT 480
+#define VREG_VSEL VREG_VOLTAGE_1_10
+#define DVI_TIMING dvi_timing_640x480p_60hz
+
+#define LED_PIN 25
```
Comment (Wren6991) [inline software/apps/mandel-full/main.c:32]: Maybe better to use PICO_DEFAULT_LED_PIN (it's not pin 25 on my DVI boards for example), but happy for it to be left as is until I sort out proper board headers

File: software/apps/mandel-full/main.c [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+      multicore_fifo_push_blocking((uint32_t)their_tmds_buf);
+  
+      queue_remove_blocking_u32(&dvi0.q_tmds_free, &our_tmds_buf);
+      absolute_time_t start_time = get_absolute_time();
+      tmds_encode_palette_data((const uint32_t*)(&mandel[y*FRAME_WIDTH]), tmds_palette, our_tmds_buf, FRAME_WIDTH, PALETTE_BITS);
+      encode_time += absolute_time_diff_us(start_time, get_absolute_time());
+      
+      multicore_fifo_pop_blocking();
+
+      while (!fractal.done && queue_get_level(&dvi0.q_tmds_valid) >= 5) generate_one_forward(&fractal);
```
Comment (Wren6991) [inline software/apps/mandel-full/main.c:202]: Ah it's based on buffer health, I was wondering how you were doing this. I guess TMDS buffers are cheaper than they used to be :)

**Review by Wren6991 — COMMENTED** [review]
Looks great, I have this running at my desk now. Please just clean up the magic numbers in tmds_encode.c, the rest will get picked up by later cleanup passes (if I get round to them)

File: software/libdvi/tmds_encode.c (hunk 2 above)
Comment (MichaelBell) [reply software/libdvi/tmds_encode.c:201]: Ah yes, these evolved from the ones in vista, I'll fix them up.

File: software/apps/mandel-full/main.c (hunk 4 above)
Comment (MichaelBell) [reply software/apps/mandel-full/main.c:32]: Cool, yep I'll fix that too.

**Comment by MichaelBell** [comment]
Thanks!  I've tidied up the magic numbers using the register constants and switched to using DEFAULT_LED_PIN.

**Pull request merged**